pyrene/src/utils/DateUtils.ts

```diff
@@ -0,0 +1,258 @@
+    }
+    return new Date(date.year, date.month - 1, date.day);
+  }
+  return undefined;
+};
+
+/**
+ * Converts a JavaScript Date object to our custom date object format
+ * Increases the month number by 1 so that it is 1-indexed
+ *
+ * @param {Date} date
+ * @returns {DateType}
+ */
+export const convertToDateTypeObject = (date: Date) : DateType => ({
+  year: date.getFullYear(),
+  month: date.getMonth() + 1,
+  day: date.getDate(),
+});
+
+/**
+ * Returns our custom time object format from a JavaScript Date object
+ *
+ * @param {Date} date
+ * @returns {TimeType}
+ */
+export const convertToTimeTypeObject = (date: Date) : TimeType => ({
+  hours: date.getHours(),
+  minutes: date.getMinutes(),
+});
+
+/**
+ * Converts custom date and time object to unix timestamp
+ *
+ * @param {DateType} date
+ * @param {TimeType} time
+ * @returns {DateType}
+ */
+export const convertToTimeStamp = (date: DateType, time: TimeType): number => {
+  // Month shift : JS Date uses 0 - 11 to count months
+  const tStamp = new Date(date.year, date.month - 1, date.day, time.hours, time.minutes);
+  return tStamp.valueOf();
+};
+
+/**
+ * Provides the Date Object of current date/time
+ */
+export const getCurrentJsDateObject = (): Date => new Date();
+
+/**
+ * Provides the DateType object of current date/time
+ */
+export const getCurrentDateTypeObject = (): DateType => convertToDateTypeObject(new Date());
+
+/**
+ * Returns the Date Object of provided timestamp in provided timezone
+ */
+export const convertToZoneTime = (timestamp: number, timezone: string): Date => utcToZonedTime(timestamp, timezone);
+
+/**
+ * Returns the Date Object of provided datetime parameter in UTC
+ */
+export const convertToUTCtime = (datetime: string | number | Date, timezone: string): Date => zonedTimeToUtc(datetime, timezone);
+
+/**
+ * Returns the timestamp of a point in time in the future relative to now
+ *
+ * @param {Duration} duration
+ */
+export const getFutureDate = (duration: Duration): number => add(getCurrentJsDateObject(), duration).valueOf();
+
+/**
+ * Returns the timestamp of a point in time in the past relative to now
+ *
+ * @param {Duration} duration
+ */
+export const getPastDate = (duration: Duration): number => sub(getCurrentJsDateObject(), duration).valueOf();
+
+/** Checks if a timezone string is valid
+* @param {string} timezone
+*/
+export const isValidTimeZone = (timezone: string): boolean => {
+  try {
+    utcToZonedTime(getCurrentJsDateObject(), timezone);
+    return true;
+  } catch {
+    return false;
+  }
+};
+
+// isExists uses 0 indexed month numbers
+/**
+ * Checks if a timetype is valid
+ * @param {DateType} date
+ */
+export const isValidDate = (date?: DateType): boolean => {
+  if (date) {
+    return isExists(date.year, date.month - 1, date.day);
+  }
+  return false;
+};
+
+/**
+ * Checks if a timetype is valid
+ * @param {TimeType} time
+ */
+export const isValidTime = (time?: TimeType): boolean => {
```

Comment:
This true but then we would not be able to do "fancy"error checking😅.  date-fns does not have the capability to tell you if it is the time part of the datetime that is wrong or the date part. You just get one error. So if we rely on the error from date-fns it would be impossible to achieve this:
![Screenshot 2021-11-05 at 11 30 18](https://user-images.githubusercontent.com/12200093/140497123-a5618bf9-5cc0-4fc9-9742-2b7b725b9c42.png)
If you have a different approach please let me know✨

Reply:
To achieve this we will need to split the date and time. There is a point though, that we may not need to have a 3rd or 4th type and just rely on the date-fns functions to tell if the date or time string is correct